Admin Config Search

maximgubar opened the pull request:
### Description
This feature will provide a possibility to find the config fields in Magento in a fast and easy way (e.g. live search).

### Manual testing scenarios
1. Open Magento admin in your browser.
2. Type the keyword in global search input field.
3. If there is a match in the config elements - results should appear in 'autocomplete results' area.
4. Each result is clickable and should redirect to you to proper section/group of configuration UI.

### Contribution checklist
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds on Travis CI are green)

maximgubar committed:
initial config search feature commit

maximgubar committed:
PR changes we applied:
- reduced text info in search results
- injection of dependency was changed to optional (backward compatibility)
- highlighting of found field
- defined vertical size of results list

maghamed commented on app/code/Magento/Backend/Model/Search/Config.php:
```diff
@@ -0,0 +1,81 @@
+     * @param string $needle
+     * @param string $pathLabel
+     */
+    public function findInStructure(ElementIterator $structureElementIterator, $needle, $pathLabel = '')
```
why does this method declared as public? 
We use it just inside this class

maghamed replied:
don't think "needle" is a good name.
Unless we call $structureElementIterator as haystack

It's the user's search term, so let's call it this way

maghamed commented on app/code/Magento/Backend/Model/Search/Config.php:
```diff
@@ -0,0 +1,81 @@
+        foreach ($structureElementIterator as $structureElement) {
+            if (!($structureElement instanceof ElementInterface)) {
+                continue;
+            }
```
why do we have this check? 

Magento\Config\Model\Config\Structure\Element\Iterator returns  \Magento\Config\Model\Config\Structure\ElementInterface by contract, do we have any reasons not to trust the declared contract and make additional check? 
```php
    /**
     * Return the current element
     *
     * @return \Magento\Config\Model\Config\Structure\ElementInterface
     */
    public function current()
    {
        return $this->_flyweight;
    }
```

maghamed commented on app/code/Magento/Backend/Model/Search/Config.php:
```diff
@@ -0,0 +1,81 @@
+            if (!($structureElement instanceof ElementInterface)) {
+                continue;
+            }
+            if (stripos((string)$structureElement->getLabel(), $needle) !== false) {
```
it's better to use 
mb_stripos than stripos
Magento requires        ` "ext-mbstring": "*",` to be installed

maghamed commented on app/code/Magento/Backend/Model/Search/Config.php:
```diff
@@ -0,0 +1,81 @@
+        if (!$this->hasQuery()) {
+            $this->setResults($this->resultBuilder->getAll());
+            return $this;
+        }
```
if we will move a check for the empty query into the findInStructure function, we could end-up with 
```php
    public function load()
     {
         $this->findInStructure($this->configStructure->getTabs(), $this->getQuery());
         $this->setResults($this->resultBuilder->getAll());
         return $this;
     }
```

maghamed commented on app/code/Magento/Backend/Model/Search/Config/Result/Builder.php:
```diff
@@ -0,0 +1,97 @@
+        self::STRUCTURE_ELEMENT_TYPE_SECTION,
+        self::STRUCTURE_ELEMENT_TYPE_GROUP,
+        self::STRUCTURE_ELEMENT_TYPE_FIELD,
+    ];
```
it's better not to hard code supported elements, but accept an additional parameter, which is configured through the DI.

So, if some new supported component would be added in future we could easily add this new type with DI configuration.

maghamed commented on app/code/Magento/Backend/Model/Search/Config/Result/Builder.php:
```diff
@@ -0,0 +1,97 @@
+                    'section' => $elementPathParts[0],
+                    'group'   => $elementPathParts[1],
+                    'field'   => $structureElement->getId(),
+                ];
```
it's better to introduce an interface and provide implementation for each of the type.
To have an ability to add new supported type in future. 
For now, we hardcoded existing structure elements

maghamed commented on app/code/Magento/Config/Block/System/Config/Edit.php:
```diff
@@ -41,16 +44,24 @@ class Edit extends \Magento\Backend\Block\Widget
         \Magento\Config\Model\Config\Structure $configStructure,
-        array $data = []
+        array $data = [],
+        Json $jsonSerializer = null
```
it's better to have a dependency on the interface instead?
`Magento\Framework\Serialize\SerializerInterface`

korostii replied:
As mentioned on https://github.com/magento/magento2/pull/8331#discussion_r99838100 and referenced on [#10342 (comment)](https://github.com/magento/magento2/pull/10342/files#r129688126) , `SerializerInterface` is not guaranteed to return a _JSON_ serializer in the future. And `\Magento\Config\Block\System\Config` uses it in a [method explicitly named `getConfigSearchParamsJson`](https://github.com/liip-forks/magento2/blob/94b423bf985dbd210a5eb13a43ff4dd92a8cc3ca/app/code/Magento/Config/Block/System/Config/Edit.php#L124)

maximgubar committed:
PR changes we applied:
- changed visibility of findInStructure() method + renamed its argument
- removed unnecessary verifications

maximgubar committed:
Merge remote-tracking branch 'upstream/develop' into develop

# Conflicts:
#	app/code/Magento/Backend/i18n/en_US.csv

maximgubar committed:
PR changes:
- introduced extended interface for \Magento\Config\Model\Config\Structure\ElementInterface
- moved structure element types from static dependency to DI

maximgubar committed:
Merge remote-tracking branch 'upstream/develop' into develop

maximgubar committed:
added unit tests

maghamed commented on app/code/Magento/Config/Model/Config/Structure/ElementNewInterface.php:
```diff
@@ -0,0 +1,20 @@
+/**
+ * @api
+ */
+interface ElementNewInterface extends ElementInterface
```
Don't think that naming is very good in this case ElementInterface -> ElementNewInterface

maximgubar replied:
I'll move it up and name it StructureElementInterface, wdyt ?

maghamed commented on app/code/Magento/Backend/Model/Search/Config/Result/Builder.php:
```diff
@@ -0,0 +1,81 @@
+{
+    const STRUCTURE_ELEMENT_TYPE_SECTION = 'section';
+    const STRUCTURE_ELEMENT_TYPE_GROUP   = 'group';
+    const STRUCTURE_ELEMENT_TYPE_FIELD   = 'field';
```
why do we need these constants if they not used

maghamed commented on app/code/Magento/Backend/Model/Search/Config.php:
```diff
@@ -0,0 +1,76 @@
+ *
+ * @method string|null getQuery()
+ * @method bool hasQuery()
+ * @method Config setResults(array $results)
```
it's better to have all these methods as a part of Object contract, not relying on `__call` usage

maghamed commented on app/code/Magento/Backend/Model/Search/Config.php:
```diff
@@ -0,0 +1,78 @@
+ * @method string|null getQuery()
+ * @method bool hasQuery()
+ * @method Config setResults(array $results)
+ * @method array getResults()
```
it's better to have these methods in class contract, not relying on `__call` method

maximgubar committed:
added integration test

maximgubar committed:
renamed ElementNewInterface to StructureElementInterface

maximgubar committed:
fixed static tests

maximgubar committed:
fixed integration tests

maximgubar committed:
updated test config for integration test

maximgubar committed:
Merge remote-tracking branch 'upstream/develop' into develop

# Conflicts:
#	app/code/Magento/Config/Model/Config/Structure/ElementInterface.php

ishakhsuvarov committed:
Merge branch 'develop' of github.com:magento/magento2 into maximgubar-develop

ishakhsuvarov commented:
Hi @maximgubar 
Please check the `Magento\Backend\Model\Search\ConfigTest` failures. Looks like search result in the test is always empty. Let us know if you need any assistance.
Thank you.

maximgubar committed:
fixed integration test

maximgubar committed:
Merge remote-tracking branch 'origin/develop' into develop

ishakhsuvarov commented:
@maximgubar Thank you for the update

maximgubar committed:
fixed static test

maximgubar committed:
fixed static test

maximgubar committed:
fixed static test

vrann commented:
@maximgubar 

- [ ] add description to the pull request
- [ ] please fix static test:
```
Magento\Test\Php\XssPhtmlTemplateTest::testXssSensitiveOutput
Passed: 1172, Failed: 1, Incomplete: 0, Skipped: 0.
Data set: /opt/bamboo/agents/agent1/xml-data/build-dir/M2-AT2472-SLE1/magento2/app/code/Magento/Config/view/adminhtml/templates/system/config/edit.phtml
Potentially XSS vulnerability. Please verify that output is escaped at lines 391
Failed asserting that a string is empty.
/opt/bamboo/agents/agent1/xml-data/build-dir/M2-AT2472-SLE1/magento2/dev/tests/static/testsuite/Magento/Test/Php/XssPhtmlTemplateTest.php:48
/opt/bamboo/agents/agent1/xml-data/build-dir/M2-AT2472-SLE1/magento2/lib/internal/Magento/Framework/App/Utility/AggregateInvoker.php:56
/opt/bamboo/agents/agent1/xml-data/build-dir/M2-AT2472-SLE1/magento2/dev/tests/static/testsuite/Magento/Test/Php/XssPhtmlTemplateTest.php:51
(4 more lines...)
```

maximgubar committed:
fixed internal static test

vrann commented:
added changes per UX request:
1. `.highlighted` should be added to tr around the input
2. `.highlighted` color should be #DFF7FF
